.ci/pitest.groovy

```diff
@@ -0,0 +1,408 @@
+        generatePitestReport(argument)
+        if (!(flag == null || (flag == "-g" || flag == "--generate-suppression"))) {
+            final String exceptionMessage = "\nUnexpected flag: ${flag}" + usageString
+            throw new IllegalArgumentException(exceptionMessage)
+        }
+        exitCode = checkPitestReport(argument, flag)
+
+    }
+    else if (argument == "--list") {
+        println "Supported profiles:"
+        profiles.each { println it }
+        exitCode = 0
+    }
+    else {
+        final String exceptionMessage = "\nUnexpected argument: ${argument}" + usageString
+        throw new IllegalArgumentException(exceptionMessage)
+    }
+    return exitCode
+}
+
+/**
+ * Parse the pom.xml file to get all the available pitest profiles.
+ *
+ * @return A set of all available pitest profiles
+ */
+private static Set<String> getPitestProfiles() {
+    final GPathResult mainNode = new XmlSlurper().parse(".${SEPARATOR}pom.xml")
+    final NodeChildren ids = mainNode.profiles.profile.id as NodeChildren
+    final Set<String> profiles = new HashSet<>()
+    ids.each { node ->
+        final GPathResult id = node as GPathResult
+        final String idText = id.text()
+        if (idText.startsWith("pitest-")) {
+            profiles.add(idText)
+        }
+    }
+    return profiles
+}
+
+/**
+ * Generate the pitest report by running the respective maven command.
+ *
+ * @param profile the pitest profile to execute
+ */
+private static void generatePitestReport(String profile) {
+    final String mavenCommand = """
+        mvn --no-transfer-progress -e -P"${profile}" 
+        clean test-compile org.pitest:pitest-maven:mutationCoverage
+        """.stripIndent()
+    final Process proc = mavenCommand.execute()
+    proc.in.eachLine { println it }
+}
+
+/**
+ * Check the generated pitest report. Parse the surviving and suppressed mutations and compare
+ * them.
+ *
+ * @param profile the pitest profile to execute
+ * @param flag command line argument flag to determine output format
+ * @return {@code 0} if pitest report is as expected, {@code 1} otherwise
+ */
+private static int checkPitestReport(String profile, String flag) {
+    final XmlParser xmlParser = new XmlParser()
+    File mutationReportFile = null
+    final String suppressedMutationFileUri =
+            ".${SEPARATOR}.ci${SEPARATOR}pitest-suppressions${SEPARATOR}${profile}-suppressions.xml"
+    new File(".${SEPARATOR}target${SEPARATOR}pit-reports")
+            .eachFileRecurse(FileType.FILES) {
+                if (it.name == 'mutations.xml') {
+                    mutationReportFile = it
+                }
+            }
+    final Node mutationReportNode = xmlParser.parse(mutationReportFile)
+    final Set<Mutation> survivingMutations = getSurvivingMutationList(mutationReportNode)
+
+    final File suppressionFile = new File(suppressedMutationFileUri)
+    Set<Mutation> suppressedMutations = Collections.<Mutation> emptySet()
```

Comment:
Is groovy not able to infer type?

Reply:
It is able to but I mentioned explicitly here, should I use def for all types except primitives and String?